velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h

```diff
@@ -0,0 +1,342 @@
+                       .singleAggregation(
+                           groupingKeysAndY, {makeArrayAggCall(aggregate)})
+                       .window({fmt::format(
+                           "row_number() over ({}) as row_number", frame)})
+                       .filter(fmt::format("row_number <= {}", n - 1))
+                       .project(projectColumnsBeforeUnion)
+                       .singleAggregation(
+                           groupingKeys, {"set_union(expected) as expected"})
+                       .project(finalProjectColumns);
+      subset_ = AssertQueryBuilder(subsetPlan.planNode())
+                    .copyResults(input[0]->pool());
+    }
+  }
+
+  bool compare(
+      const RowVectorPtr& /*result*/,
+      const RowVectorPtr& /*altResult*/) override {
+    VELOX_UNSUPPORTED();
+  }
+
+  bool verify(const RowVectorPtr& result) override {
+    // Union 'result' with 'superset_', group by on 'groupingKeys_' and produce
+    // pairs of actual and expected values per group. We cannot use join because
+    // grouping keys may have nulls.
+    auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+    auto expectedSource =
+        PlanBuilder(planNodeIdGenerator).values({superset_}).planNode();
+
+    // Append a NULL column for expected so that we can union 'result' with
+    // 'superset_' and 'subset_'.
+    std::string nullExpected;
+    if (minMaxByN_) {
+      nullExpected =
+          fmt::format("cast(NULL as {}) as expected", aggregateTypeSql_);
+    } else {
+      nullExpected =
+          fmt::format("cast(NULL as {}[]) as expected", aggregateTypeSql_);
+    }
+    auto actualSource = PlanBuilder(planNodeIdGenerator, result->pool())
+                            .values({result})
+                            .appendColumns({nullExpected})
+                            .planNode();
+
+    core::PlanNodePtr plan;
+    if (minMaxByN_) {
+      plan =
+          PlanBuilder(planNodeIdGenerator, result->pool())
+              .localPartition({}, {expectedSource, actualSource})
+              .singleAggregation(
+                  groupingKeys_,
+                  {fmt::format("array_agg({}) as a", name_),
+                   "array_agg(expected) as e"})
+              .project({"remove_nulls(a) as a", "remove_nulls(e) as e"})
+              // Check that if actual result is not NULL, all elements in a[1]
+              // are also in e[1]. If the actual result is NULL, either the
+              // expected result is NULL too (when all rows in the group are
+              // masked out), or e[1] is empty (when all y in the group are
+              // NULL).
+              .project({fmt::format(
+                  "switch(cardinality(a) > 0, cardinality(e) > 0 and reduce(a[1], 0, (s, x) -> if (s < 0, -1, if (x is null, if (cardinality(e[1]) > cardinality(remove_nulls(e[1])), 1, -1), if (\"$internal$contains\"(e[1], x), 1, -1))), s -> s >= 0), cardinality(e) = 0 or cardinality(e[1]) = 0)",
```

Comment:
Its pretty hard to parse this switch/reduce , wonder if theres a way to simplify (or if not a better explanation than the comment above).

Reply:
Extracted the common logic in the update. Thanks!